File: Tree/Path_Sum_III.py
#Problem Link :- https://leetcode.com/problems/path-sum-iii/

"""You are given a binary tree in which each node contains an integer value.Find the 
number of paths that sum to a given value.The path does not need to start or end at 
the root or a leaf, but it must go downwards (traveling only from parent nodes to 
child nodes)."""

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        d = dict()
        d[0] = 1
        self.count = 0
        def helper(root,ps):
            if root is None:
                return 0
            cs = ps + root.val
            x = cs - sum
            if x in d:
                self.count += d[x]
            if cs in d:
                d[cs] += 1
            else:
                d[cs] = 1
            helper(root.left,cs)
            helper(root.right,cs)
            d[cs] -= 1
        helper(root,0)
        return self.count
            
            

File: Tree/PathSumIII.cpp
//Problem Link:- https://leetcode.com/problems/path-sum-iii/

#include<bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int res = 0;
    void helper(TreeNode* root,int sm,unordered_map<int,int>&mp,int target){
        if (!root){
            return;
        }
        sm = sm + root->val;
        if (mp.find(sm-target) != mp.end()){
            res += mp[sm-target];
            mp[sm]++;
        }
        else{
            mp[sm]++;
        }
        helper(root->left,sm,mp,target);
        helper(root->right,sm,mp,target);
        mp[sm]--;
        if (mp[sm] == 0){
            mp.erase(sm);
        }
    }
    
    int pathSum(TreeNode* root, int target) {
        if (!root){
            return 0;
        }
        if (!root->left && !root->right && root->val == target){
            return 1;
        }
        unordered_map<int,int> mp;
        mp[0] =  1;
        helper(root,0,mp,target);
        return res;
    }
};